resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java

```diff
@@ -1133,15 +1146,58 @@ public void run() {
+                                       final DnsCache resolveCache,
+                                       final boolean completeEarlyIfPossible) {
 
         assert executor().inEventLoop();
 
+        if (inflightLookups != null) {
+            if (additionals == null || additionals.length == 0) {
+                Future<List<InetAddress>> inflightFuture = inflightLookups.get(hostname);
+                if (inflightFuture != null) {
+                    inflightFuture.addListener(new GenericFutureListener<Future<? super List<InetAddress>>>() {
+                        @SuppressWarnings("unchecked")
+                        @Override
+                        public void operationComplete(Future<? super List<InetAddress>> future) {
+                            if (future.isSuccess()) {
+                                promise.setSuccess((List<InetAddress>) future.getNow());
+                            } else {
+                                Throwable cause = future.cause();
+                                if (isTimeoutError(cause)) {
+                                    // The failure was caused by a timeout. This might be happening as a result of
+                                    // the remote server to be overloaded for some short amount of time or because
+                                    // UDP packets were dropped on the floor. In this case lets try to just do the
+                                    // query explicit to not cascade this possible temporary failure.
+                                    final DnsServerAddressStream nameServerAddrs =
+                                            dnsServerAddressStreamProvider.nameServerAddressStream(hostname);
+                                    new DnsAddressResolveContext(DnsNameResolver.this, originalPromise,
+                                            hostname, additionals, nameServerAddrs,
+                                            maxQueriesPerResolve, resolveCache,
+                                            authoritativeDnsServerCache, completeEarlyIfPossible)
+                                            .resolve(promise);
+                                } else {
+                                    promise.setFailure(cause);
+                                }
+                            }
+                        }
+                    });
+                    return;
+                // Check if we have space left in the map.
+                } else if (maxNumConsolidation == 0 || inflightLookups.size() < maxNumConsolidation) {
+                    inflightLookups.put(hostname, promise);
+                    promise.addListener(new GenericFutureListener<Future<? super List<InetAddress>>>() {
+                        @Override
+                        public void operationComplete(Future<? super List<InetAddress>> future) throws Exception {
+                            inflightLookups.remove(hostname);
```

Comment:
do we need to worry about re-entry? if future notification order is guaranteed to be ordered so that the "original" future listener will be notified then non issue but otherwise what if the completion of a DNS resolution triggers another DNS resolution, futures that complete from the first resolution may remove entries for new inflight requests iiic?